core/matrix/ell.hpp

```diff
@@ -0,0 +1,268 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright 2017-2018
+
+Karlsruhe Institute of Technology
+Universitat Jaume I
+University of Tennessee
+
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its contributors
+   may be used to endorse or promote products derived from this software
+   without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_MATRIX_ELL_HPP_
+#define GKO_CORE_MATRIX_ELL_HPP_
+
+
+#include "core/base/array.hpp"
+#include "core/base/convertible.hpp"
+#include "core/base/lin_op.hpp"
+#include "core/base/mtx_reader.hpp"
+
+
+namespace gko {
+namespace matrix {
+
+template <typename ValueType>
+class Dense;
+
+/**
+ * Ell is a matrix format which aligned all nonzero elements to the left
```

Comment:
ELL is a matrix format where padding with explicit zeros is used such that all rows have the same number of stored elements.  The number of elements stored in each row is the largest number of nonzero elements in any of the rows (parameter max_nnz_row).  This removes the need of a row pointer like in the CSR format, and allows for SIMD processing of the distinct rows.   For efficient processing, the nonzero elements and the corresponding column indices are stored in column-major fashion. The columns are padded to a multiple of 32 for efficient memory access.

Reply:
We commented on this at the same time. We should somehow merge our proposals. In general, I like yours better than mine, but there is some useful information in both of our proposals which is not present in the other. I would only change:

> (parameter max_nnz_row)

to:

> (obtainable through get_max_nonzeros_per_row() method)

Reply:
ok

Reply:
Modified